pkg/acl/acl.go

```diff
@@ -0,0 +1,230 @@
+/*
+ * JuiceFS, Copyright 2024 Juicedata, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package acl
+
+import (
+	"fmt"
+	"hash/crc32"
+	"sort"
+	"strings"
+
+	"github.com/juicedata/juicefs/pkg/utils"
+)
+
+const Version uint8 = 2
+
+type Entry struct {
+	Id   uint32
+	Perm uint16
+}
+
+type Entries []Entry
+
+func (es *Entries) Len() int           { return len(*es) }
+func (es *Entries) Less(i, j int) bool { return (*es)[i].Id < (*es)[j].Id }
+func (es *Entries) Swap(i, j int)      { (*es)[i], (*es)[j] = (*es)[j], (*es)[i] }
+
+func (es *Entries) IsEqual(other *Entries) bool {
+	if es.Len() != other.Len() {
+		return false
+	}
+	sort.Sort(es)
+	sort.Sort(other)
+	for i := 0; i < es.Len(); i++ {
+		if (*es)[i].Id != (*other)[i].Id || (*es)[i].Perm != (*other)[i].Perm {
+			return false
+		}
+	}
+	return true
+}
+
+func (es *Entries) String() string {
+	var builder strings.Builder
+	builder.WriteString("{")
+	for _, e := range *es {
+		builder.WriteString(fmt.Sprintf("{id: %d, perm: %o}, ", e.Id, e.Perm))
+	}
+	builder.WriteString("}")
+	return builder.String()
+}
+
+func (es *Entries) Encode() []byte {
+	w := utils.NewBuffer(4 + uint32(es.Len()*6))
+	for _, e := range *es {
+		w.Put32(e.Id)
+		w.Put16(e.Perm)
+	}
+	return w.Bytes()
+}
+
+func (es *Entries) Decode(data []byte) {
+	r := utils.ReadBuffer(data)
+	for r.HasMore() {
+		*es = append(*es, Entry{
+			Id:   r.Get32(),
+			Perm: r.Get16(),
+		})
+	}
+}
+
+type Rule struct {
+	Owner       uint16
+	Group       uint16
+	Mask        uint16
+	Other       uint16
+	NamedUsers  Entries
+	NamedGroups Entries
+}
+
+func (r *Rule) String() string {
+	return fmt.Sprintf("owner %o, group %o, mask %o, other %o, named users: %s, named group %s",
+		r.Owner, r.Group, r.Mask, r.Other, r.NamedUsers.String(), r.NamedGroups.String())
+}
+
+func (r *Rule) Encode() []byte {
+	w := utils.NewBuffer(uint32(16 + (len(r.NamedUsers)+len(r.NamedGroups))*6))
+	w.Put16(r.Owner)
+	w.Put16(r.Group)
+	w.Put16(r.Mask)
+	w.Put16(r.Other)
+	w.Put32(uint32(len(r.NamedUsers)))
+	for _, entry := range r.NamedUsers {
+		w.Put32(entry.Id)
+		w.Put16(entry.Perm)
+	}
+	w.Put32(uint32(len(r.NamedGroups)))
+	for _, entry := range r.NamedGroups {
+		w.Put32(entry.Id)
+		w.Put16(entry.Perm)
+	}
+	return w.Bytes()
+}
+
+func (r *Rule) Decode(buf []byte) {
+	rb := utils.ReadBuffer(buf)
+	r.Owner = rb.Get16()
+	r.Group = rb.Get16()
+	r.Mask = rb.Get16()
+	r.Other = rb.Get16()
+	uCnt := rb.Get32()
+	if uCnt != 0 {
```

Comment:
it's OK to `make([]Entry, 0)`

Reply:
default is nil, no need to allocate an empty slice

Reply:
It's better to remove the `if`, less branches is better.